### PR TITLE
Implement language feature #6885

### PR DESCRIPTION
**Pull request by cooldome**
The idea is simple: ignore always raising exceptions branches in if/case exprs, so return deducted type can be not nil. The following now compiles:
```let str1 = "Y"
  let a = case str1:
    of "Y": true
    of "N": false
    else: raise newException(ValueError, "Invalid boolean")
```
Type of a variable is bool, exception raising branch is ignored. I don't expect any backwards in compatible behavior, as previously such code didn't compile.

Implementation detail: raise stmt tyVoid tag was selected at random, to be able to detect it in commonType() proc. It could be another type tag, let me if there are better alternatives.

Tests added to cover cases that should not compile and several other that should not compile

### CODE REVIEW

**Comment by cooldome** [comment]
I see failed tests, I have an idea to fix it

File: compiler/sem.nim [hunk 1]
```diff
@@ -105,8 +105,8 @@ proc commonType*(x, y: PType): PType =
   var a = skipTypes(x, {tyGenericInst, tyAlias})
   var b = skipTypes(y, {tyGenericInst, tyAlias})
   result = x
-  if b.kind in {tyExpr, tyNil, tyVoid}: result = x
-  elif a.kind in {tyExpr, tyNil}: result = y
+  if a.kind in {tyExpr, tyNil}: result = y
```
Comment (Araq) [inline compiler/sem.nim]: Why swap these two tests?

Comment (cooldome) [reply compiler/sem.nim]: in the latest version, they are no longer swapped. In first version, I was using tyVoid tag on raise statement and needed to check b for tyVoid before a. In latest version, raise statement has no type as originally, I am checking for node kinds instead, a and b checks are reverted to the original order

File: compiler/sem.nim [hunk 2]
```diff
@@ -165,6 +165,11 @@ proc commonType*(x, y: PType): PType =
         result = newType(k, r.owner)
         result.addSonSkipIntLit(r)
 
+proc commonType*(x: PType, y: PNode): PType =
+  # ignore exception raising branches in case/if expressions
+  if y.kind == nkRaiseStmt: return x
```
Comment (Araq) [inline compiler/sem.nim]: This is very good, but not completely correct. You need to do something like

```nim

proc endsInRaise(n: PNode): bool =
  var it = n
  while it.kind in {nkStmtList, nkStmtListExpr}: it = it.lastSon
  result = it.kind == nkRaiseStmt
```

**Comment by cooldome** [comment]
The following issues should be closed now: #6885 #842

**Comment by Yardanico** [comment]
@cooldome closed them

**Comment by dom96** [comment]
Issues shouldn't be closed until the PRs get merged. This hasn't been merged yet.

**Comment by Yardanico** [comment]
Oh, I thought this was closed

File: compiler/sem.nim [hunk 3]
```diff
@@ -165,6 +165,19 @@ proc commonType*(x, y: PType): PType =
         result = newType(k, r.owner)
         result.addSonSkipIntLit(r)
 
+proc endsInNoReturn(n: PNode): bool =
+  # check if expr ends in raise exception or call of noreturn proc
+  var it = n
+  while it.kind in {nkStmtList, nkStmtListExpr}:
```
Comment (Araq) [inline compiler/sem.nim]: ``and it.len > 0``

File: compiler/sem.nim [hunk 4]
```diff
@@ -165,6 +165,19 @@ proc commonType*(x, y: PType): PType =
         result = newType(k, r.owner)
         result.addSonSkipIntLit(r)
 
+proc endsInNoReturn(n: PNode): bool =
+  # check if expr ends in raise exception or call of noreturn proc
+  var it = n
+  while it.kind in {nkStmtList, nkStmtListExpr}: 
+    it = it.lastSon
+  result = it.kind in nkCallKinds and sfNoReturn in it[0].sym.flags
+  result = result or it.kind == nkRaiseStmt
```
Comment (Araq) [inline compiler/sem.nim]: Merge these two ``result`` assignments into one.

**Comment by cooldome** [comment]
One more attempt, more cases covered

**Comment by cooldome** [comment]
Just to let you know I have one more pull request that is going to use procSymInCallExpr(): disallow statements after noreturn proc call, just like statements after break, continue and return are disallowed

**Comment by cooldome** [comment]
One more thing. endsInNoReturn is called twice for every branch in if/case expressions. I can make it only a single call per branch and store intermediate result in seq[bool] and reuse second time, but I am not sure if it is going to make any difference. Let me know if you think, it is worth it optimization

File: compiler/sem.nim [hunk 5]
```diff
@@ -165,6 +165,27 @@ proc commonType*(x, y: PType): PType =
         result = newType(k, r.owner)
         result.addSonSkipIntLit(r)
 
+proc procSymInCallExpr(n: PNode): PSym =
+  assert(n.kind in nkCallKinds)
+  var it = n
+  while it.kind != nkSym:
```
Comment (Araq) [inline compiler/sem.nim]: That is not correct and not necessary!  Should produce a crashing loop for ``f()()``. Proc calls are not static in Nim, they can always go through a procvar!

File: compiler/sem.nim [hunk 6]
```diff
@@ -165,6 +165,27 @@ proc commonType*(x, y: PType): PType =
...
+         else: it[0]
+  it.sym
+
+proc endsInNoReturn(n: PNode): bool =
+  # check if expr ends in raise exception or call of noreturn proc
+  var it = n
+  while it.kind in {nkStmtList, nkStmtListExpr} and it.len > 0: 
+    it = it.lastSon
+  result = it.kind == nkRaiseStmt or 
+    (it.kind in nkCallKinds and sfNoReturn in procSymInCallExpr(it).flags)
```
Comment (Araq) [inline compiler/sem.nim]: The correct check is ``it.kind in nkCallKinds and it[0].kind == nkSym and sfNoReturn in it[0].sym.flags``

File: compiler/sem.nim (hunk 5 above)
Comment (cooldome) [reply compiler/sem.nim]: Thanks for clarification, now it is clear that dotExprs that I was seeing were calls via function pointers and should be ignored.
procSymInCallExpr is removed.

File: compiler/sem.nim (hunk 6 above)
Comment (cooldome) [reply compiler/sem.nim]: done, as you say

**Comment by cooldome** [comment]
Add more sematics checks for noreturns procs: return type is not allowed for noreturn procs, statements after call to noreturn procs are no longer allowed.

File: compiler/pragmas.nim [hunk 7]
```diff
@@ -771,6 +771,8 @@ proc singlePragma(c: PContext, sym: PSym, n: PNode, i: int,
       of wNoreturn:
         noVal(it)
         incl(sym.flags, sfNoReturn)
+        if sym.ast[paramsPos][0].kind != nkEmpty:
+          localError(sym.ast[paramsPos][0].info, errNoReturnWithReturnTypeNotAllowed)
```
Comment (Araq) [inline compiler/pragmas.nim]: I'll merge this now anyway, but this check is not 100% correct, ``void`` as a return type would be valid.

**Pull request merged**